### ISSUE TITLE
Transactions are not guaranteed to run according to their arrival order

### ISSUE THREAD

**Issue by m-Peter** [onflow/flow-evm-gateway#699]
Currently, for each transaction submitted with `eth_sendRawTransaction`, we wrap an `EVM.run` call within a Flow transaction, sign it, and submit it in the Flow network using an Access Node.

If 2 transactions A, B are submitted with a very slight delay (say milliseconds), there's a chance that the Flow transactions that wrap EVM transactions A, B get executed in a reverse order. This is due to the pipeline architecture of Flow. The Flow transactions for A, B might be hashed to different collections, and the consensus node might choose to run B before A.

For the most part, transactions A, B might be unrelated to each other, but if they originate from the same EOA, then transaction B will fail with a `nonce too high` error, while transaction A will succeed.. There might be other cases where the reversal of transaction execution affects the end result.

One possible solution to this, is to have a TxPool, that will accept incoming transactions, and use `EVM.batchRun`, to execute them according to their arrival order. In order for this to work properly, we'll have to fine tune the amount of period a transaction should stay in the TxPool, before being submitted with `EVM.batchRun`.

Related issue: https://github.com/onflow/flow-evm-gateway/issues/477

Appendix: An example of a Flow block on testnet, where there's 3 failing EVM transactions from the same EOA: https://testnet.flowscan.io/block/231321798.

**Comment by bluesign** [comment]
I am not sure if `batchRun` will fix this. is it panic safe?

**Comment by m-Peter** [comment]
> I am not sure if `batchRun` will fix this. is it panic safe?

@bluesign Here you can see a usage of `EVM.batchRun` in action:
```cadence
transaction(encodedTxs: [String]) {
    prepare(signer: auth(Storage) &Account) {
        var txs: [[UInt8]] = []
        for enc in encodedTxs {
            txs.append(enc.decodeHex())
        }

        let txResults = EVM.batchRun(
            txs: txs,
            coinbase: EVM.EVMAddress(bytes: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19])
        )
        for txResult in txResults {
            assert(txResult.status == EVM.Status.successful, message: "failed to execute tx, error: ".concat(txResult.errorCode.toString()))
        }
    }
}
```
It accepts an array of RLP-encoded transactions, executes them in their given order, and returns the execution result. It does not panic or anything like that, it allows the caller to decide what to do, based on the result of each executed transaction.
Do you have any specific concern that this approach is not going to fix the tx ordering issue?

**Comment by bluesign** [comment]
@m-Peter  sorry I was too brief in first message, what I meant was; what if batchRun rans out of computation and panics. 

afaik there is no direct relationship with gas and cadence computation, how we will batch the transactions ? Even if there was direct relationship, can't someone send a very high gas limit transaction, but only pay for what is executed?

**Comment by janezpodhostnik** [comment]
> Something like 3000 EVM gas for 1 Cadence computation.

That is correct.

Here is an example on how it works:
- lets say FVM is executing a cadence transaction with computation limit 1000
- The transaction used up 150 computation so far
- An EVM transaction is encountered with a gas limit of 300.000
- FVM checks if there is enough computation left `1000 - 150 > 300.000 / 3000` `850 > 100` 
- because there is it gets executed but only 30.000 gas was used. so only 10 computation is deducted.
- The remaining computation is `850 - 10 = 840`

**Comment by bluesign** [comment]
> because there is it gets executed but only 30.000 gas was used. so only 10 computation is deducted.
The remaining computation is 850 - 10 = 840

I don't think this is correct considering storage read and writes. ( I complained on this too many times, let me explain a bit ) 

Here I have 2 successful  gateway transactions, which are pretty similar except the gas usage. 


let's take this transaction as example: ( which has 151 execution effort ) 
https://www.flowscan.io/tx/1af3d88333e1971ca4ca4c55be8d4fb996afe6a7cbd1133d0f249568995e46e9

This has 2 EVM transactions:  ( one for main transaction ( 38013 gas ) , and 21000 gas for coinbase transfer )

So 38013 + 21000 = 59013 total gas used ( this is predictable as we have coinbase transfer always 21000 ) 

59013 / 30000 = 1.967  ~= 2 computation 

now this has overhead of ~150 computation  ( which matches hypothetical example from Janez ) 


now let's another example with higher execution effort: ( which has 356 execution effort ) 

https://www.flowscan.io/tx/d3b63de19c2f7d7dbd56a806d8a9acebc81d5885a5d2de8a809b20078e1a098e

120913 + 21000 = 141913 total gas used for this one 

141913 / 30000 = 4.73  ~= 5 computation

Now we would expect this would cost ~153 computation, but now this has difference of ~203 computation  ( difference is more than our prediction ) 

So gas usage cost is translating to Flow computation effort: 

you can simply check with opcode costs also:

SSTORE non-zero to non-zero ( most common probably )  ~= 5000 evm gas = 5000 / 30000 ~= 0.16 cadence computation ( which gives us around ~13 bytes of storage read/write ) how atree with inlining works here has much more cost. 

I think when adjusting execution efforts in the future, EVM also will need some work to make it more predictable. 

There were edge cases when initial computation limit check success but EVM.run can run out of computation while executing. ( when I was developing COA Attachment I hit this before )


PS:  technically now transactions on Flow are almost free and this is not an issue for now.

**Comment by bluesign** [comment]
> One possible solution to this, is to have a TxPool, that will accept incoming transactions, and use EVM.batchRun, to execute them according to their arrival order. In order for this to work properly, we'll have to fine tune the amount of period a transaction should stay in the TxPool, before being submitted with EVM.batchRun.

I think instead of waiting maybe we can rate limit send by account address.

**Comment by m-Peter** [comment]
Another example of transactions with same EOA origin, in the same block:
1. https://testnet.flowscan.io/tx/9519a8ac0f5ed068cd109e209c7bd8e4aeebc07c435d4617d853f61feffb5c5a
2. https://testnet.flowscan.io/tx/6b430d9b633988ea92c29e574c47cb678ce82a4482c205ad7bd711d792316b00
In this case, however, they were executed successfully. The nonces were `54981` & `54982` for 1. and 2. respectively.

**Comment by m-Peter** [comment]
> > One possible solution to this, is to have a TxPool, that will accept incoming transactions, and use EVM.batchRun, to execute them according to their arrival order. In order for this to work properly, we'll have to fine tune the amount of period a transaction should stay in the TxPool, before being submitted with EVM.batchRun.
> 
> I think instead of waiting maybe we can rate limit send by account address.

@bluesign That is a good workaround actually, as we observed that this issue has intensified lately, for testnet that is. However, it might be off-putting for some devs trying out various use-cases.

**Comment by Kay-Zee** [comment]
I think when we talk about rate limit, we can consider that to be rate limiting the speed which they exit the pool, maybe?

I.e. you can send 100tx into the gateway, but the gateway will decide at which rate they release it to the ANs. This allows users to still send many tx at once, but obviously this would slow down the rate at which they might get the results of the transactions.

**Comment by bluesign** [comment]
@Kay-Zee yeah this is exactly what I was thinking. Not on ingestion but rate limiting sending of transaction to flow network by evm adress. (Considering batching will be by evm adress)
